torch/testing/asserts.py

```diff
@@ -0,0 +1,389 @@
+    """Asserts that the values of two tensors are bitwise equal.
+
+    Args:
+        a (torch.Tensor): First tensor.
+        b (torch.Tensor): Second tensor.
+
+    Raises:
+         AssertionError: If the values of :attr:`a` and :attr:`b` are not bitwise equal.
+
+    .. seealso::
+
+        Internally :func:`torch.eq` is used to check for bitwise equality.
+    """
+    if not torch.all(torch.eq(a, b)):
+        raise AssertionError("ADDME")
+
+
+def _assert_values_allclose(
+    a: torch.Tensor, b: torch.Tensor, /, *, rtol: Optional[float] = None, atol: Optional[float] = None
+) -> None:
+    """Asserts that the values of two tensors are equal up to a desired precision.
+
+    Args:
+        a (torch.Tensor): First tensor.
+        b (torch.Tensor): Second tensor.
+        rtol (Optional[float]): Relative tolerance. If omitted, a default value based on the
+            :attr:`~torch.Tensor.dtype` is selected with the table from :func:`assert_tensors_allclose`.
+        atol (Optional[float]): Absolute tolerance. If omitted, a default value based on the
+            :attr:`~torch.Tensor.dtype` is selected with the table from :func:`assert_tensors_allclose`.
+
+    Raises:
+         AssertionError: If the values of :attr:`a` and :attr:`b` are equal up to the desired precision.
+
+    .. seealso::
+
+        Internally :func:`torch.allclose` is used to check for bitwise equality.
+    """
+    default_rtol, default_atol = _get_default_rtol_and_atol(a, b)
+    rtol = rtol if rtol is not None else default_rtol
+    atol = atol if atol is not None else default_atol
+
+    if not torch.allclose(a, b, rtol=rtol, atol=atol):
+        raise AssertionError("ADDME")
+
+
+def _compare_tensors(
+    a: torch.Tensor,
+    b: torch.Tensor,
+    /,
+    values_asserter: Callable[[torch.Tensor, torch.Tensor], None],
+    *,
+    check_device: bool = True,
+    check_dtype: bool = True,
+    check_stride: bool = True,
+) -> None:
+    """Compare values and attributes and values.
+
+    Args:
+        a (torch.Tensor): First tensor.
+        b (torch.Tensor): Second tensor.
+        values_asserter (Callable[[torch.Tensor, torch.Tensor], None]): Will be called with :attr:`a` and :attr:`b`
+            after the attribute checks. Must raise an :class:`AssertionError` if the values do not match.
+        check_device (bool): If ``True`` (default), asserts that both :attr:`a` and :attr:`b` live in the same
+            :attr:`~torch.Tensor.device` memory. If this check is disabled **and** :attr:`a` and :attr:`b` do not live
+            in the same memory :attr:`~torch.Tensor.device`, they are moved to the CPU memory.
+        check_dtype (bool): If ``True`` (default), asserts that both :attr:`a` and :attr:`b` have the same
+            :attr:`~torch.Tensor.dtype`. If this check is disabled **and** :attr:`a` and :attr:`b` do not have the same
+            :attr:`~torch.Tensor.dtype`, the comparison :attr:`~torch.Tensor.dtype` is determined by
```

Comment:
To be more parallel with the device check language, this would be that the tensors are copied to the dtype returned from promote_types() before their values are compared.

Reply:
Aside: this PR definitely doesn't need to mention this, but the actual behavior is different because it uses this:

https://github.com/pytorch/pytorch/blob/547f43576399eb4237eb005ffb91039f29bdab58/torch/testing/_internal/common_utils.py#L834

In the future we can look at actually using promote_types().